### PR TITLE
[PARTI-358] Support category metadata

### PR DESCRIPTION
**Pull request by pedro-lb**
# PR Description

This PR adds the following changes:

- Supports category metadata via a new `data` field in `item_groups`
- Adds a bunch of tests + snapshots to make sure we're building the correct csv for json stuff

This field needs to be an object. We support any level of nested properties here. Refer to [the documentation](https://docs.constructor.io/rest_api/item_groups/#add-item-group) for more details. 

## ⚠️ Dependencies

- None

## 🕹 Demo

Using the script provided in the test section below, we get the data ingested. Specifically, look at the groups section:

image

You can see that the items and variations are correct, and have the correct item groups assigned:

image

image

Finally, when fetching the item groups via API, we can see the payload that was sent:

image

# How to test

## 🧪 What should be tested?

As a reviewer, you should test:

- You should be able to ingest category data as json

## 📝 How to test?

Reproduce the following steps to test:

1. Use the following script at `dev.ts` and have fun:

```ts
import { config } from "dotenv";

import { CatalogIngestor } from "./catalogIngestor";
import { CatalogIngestionType } from "./catalogIngestor/types";

config();

async function test() {
  const ingestor = new CatalogIngestor({
    apiKey: "YOUR_API_KEY_HERE",
    apiToken: "YOUR_API_TOKEN_HERE",
    force: true,
  });

  await ingestor.ingest(async () => {
    return {
      data: {
        // insert your test data here
        groups: [
          {
            id: "All",
            name: "All",
            parent_id: null,
          },
          {
            id: "test",
            name: "test",
            parent_id: "All",
            data: {
              foo: "bar",
              baz: "jazz",
              a: {
                b: {
                  c: {
                    d: {
                      e: {
                        f: {
                          g: {
                            h: {
                              i: {
                                j: {
                                  enough: "already",
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        ],
        items: [
          {
            id: "test",
            active: true,
            description: "test",
            group_ids: ["test"],
            facets: [],
            metadata: [],
            keywords: [],
            image_url:
              "https://www.google.com/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png",
            item_name: "Test Item",
            url: "https://www.google.com",
          },
        ],
        variations: [
          {
            active: true,
            facets: [],
            metadata: [],
            image_url:
              "https://www.google.com/images/branding/googlelogo/1x/googlelogo_color_272x92dp.png",
            item_id: "test",
            item_name: "Test Item Variation",
            url: "https://www.google.com",
            variation_id: "test-variation",
          },
        ],
      },
      type: CatalogIngestionType.FULL,
    };
  });
}

void test();
```

### CODE REVIEW

**Review by pedro-lb — COMMENTED** [review]
Self review.

File: src/constructor/node/buildCsvPayload/index.spec.ts [hunk 1]
```diff
@@ -194,6 +326,61 @@ describe(buildCsvRow, () => {
...
+
+        expect(result).toContain("parent_id,id,name,data");
+        expect(result).not.toContain("json:");
+      });
+
+      describe("when identifying the root value type", () => {
+        it("should NOT include the `data` column if value is a string", () => {
+          const result = buildCsvPayloadWithGroupWithJson("just a string");
+          expect(result).not.toContain("data");
+        });
```
Comment (guiquintelas) [inline src/constructor/node/buildCsvPayload/index.spec.ts:362]: **Question**: Why would that be the case?

Strings should be considered as valid json, right?

Comment (guiquintelas) [reply src/constructor/node/buildCsvPayload/index.spec.ts:362]: The same comment for other types right under this one

File: src/constructor/node/buildCsvPayload/index.spec.ts [hunk 2]
```diff
@@ -44,17 +46,147 @@ describe("when evaluating the generated csv", () => {
...
+          ],
+        },
+      });
+      const result = await buildCsvPayload(fileHelper, data);
+
+      expect(result.groups).toBeDefined();
+      await snapshotCsv(result.groups as fs.ReadStream);
+    });
+
+    it("should generate the correct csv for items with metadata and facets", async () => {
```
Comment (guiquintelas) [inline src/constructor/node/buildCsvPayload/index.spec.ts:81]: **Suggestion**: We should also have a snapshot test for item_groups :)

**Review by guiquintelas — APPROVED** [review]
Working great! And solid implementation!!

Thanks for the work here pedro. Left some questions/suggestions but nothing blocking.

File: src/constructor/node/buildCsvPayload/index.spec.ts (hunk 1 above)
Comment (pedro-lb) [reply src/constructor/node/buildCsvPayload/index.spec.ts:362]: It is, but in our API this `data` property only accepts objects.

image

File: src/constructor/node/buildCsvPayload/index.spec.ts (hunk 2 above)
Comment (pedro-lb) [reply src/constructor/node/buildCsvPayload/index.spec.ts:81]: We do have it - it's on line 63:

image

Should we add any other conditions?

Comment (guiquintelas) [reply src/constructor/node/buildCsvPayload/index.spec.ts:81]: Ohh missed hard those - I think this is enough, no need for extra conditions

**Pull request merged**